README
Test-C2FIT version 0.01
=======================

Test::C2FIT is a direct Perl port of Ward Cunningham's FIT 
acceptance test framework for Java.

SYNOPSIS

    FileRunner.pl input_containing_fit_tests.html test_results.html

DESCRIPTION

Great software requires collaboration and communication. Fit is a tool for
enhancing collaboration in software development. It's an invaluable way
to collaborate on complicated problems - and get them right - early
in development.

Fit allows customers, testers, and programmers to learn what their 
software should do and what it does do. It automatically compares
customers' expectations to actual results.

This port of FIT has a featureset that lies somewhere between
Java versions b021021j and v1.1 of FIT.  Dave W. Smith's original port
was based on fit-b021021j and I've updated most of the core to match the
1.1 version.

This port passes the current FIT spec and also implements a subset of the
examples, including the complex MusicExample.

GOTCHAS AND LIMITATIONS

1) Java is a strongly typed language; Perl is not. The Java version of FIT
cares a lot about types, but Perl takes a more relaxed view of things and
this port reflects that.

2) Perl supports limited introspection. Because there are no method signatures,
it isn't possible to determine method return types. Without known return
types, the "right" TypeAdapters can't be set up. This can cause problems
with RowFixtures and ActionFixtures.

3) Some of the tests from the 'examples' directory expect Java behaviour for
arithmetic (e.g. integer overflow).  Perl doesn't have this type of overflow
so these tests will "fail".

4) The MusicExample uses a clock that doesn't have millisecond accuracy. This
throws off the clock by a second during one of the tests.

5) Perl supports a limited set of primitive types. Dave has used a
GenericTypeAdapter that knows about strings and numbers (and pretends
to know about booleans).


SEE ALSO

Extensive and up-to-date documentation on FIT can be found at:
http://fit.c2.com/

The 'examples' directory of this distribution contains some sample FIT
tests and sample applications that they test.  Invoke FileRunner.pl on
any of the test input files from examples/input and view the output in
a browser.

You should also examine and run the tests in the 'spec' directory.
These are FIT's own acceptance tests.

AUTHOR

Original port from the Java version by Dave W. Smith
Updates and modifications by Tony Byrne <fit4perl@byrnehq.com>
Newer Updates and modification by Martin Busik <martin.busik@busik.de>

COPYRIGHT AND LICENSE

Copyright (c) 2002-2005 Cunningham & Cunningham, Inc.
Released under the terms of the GNU General Public License version 2 or later.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Test::Unit 0.24 - Required to run the unit tests.
  Error           - Implements exception handling.

COPYRIGHT AND LICENCE

Copyright (c) 2002-2006 Cunningham & Cunningham, Inc.
Released under the terms of the GNU General Public License version 2 or later.



Changes
Revision history for Perl extension Test::C2FIT.

0.07  Sun Jun 04 2006   Martin Busik <martin.busik@busik.de>
    - all calls to subs in Test::C2FIT-Packages is now done in the
      "method" style, e.g. pkg->subname and not pkg::subname
      (as only this style allows usage of static methods in subclasses)
    - all constructor calls now pkg->new instead of new pkg()
    - error handling in all Test::C2FIT classes now consistently
      try/catch/otherwise
    - unicode-characters handling in the spec (esp. nbsp (&#0160;))
      modified. Perl creates internally (at least under 5.8.8) utf8
      byte sequences. Input files are expected to be in iso-8859-* or
      windows-1252 so these characters may be encoded by a single byte.
      Therefore, a bit unpretty handling was implemented
    - new makefile-target my_tidy - formats
    - further tests in the t directory added
0.06  Sat May 13 2006   Martin Busik <martin.busik@busik.de>
    - Makefile.PL regarding META.yml modified
    - fit_shell, an interactive shell for easy running the tests
      added.
    - startet with pod documentation
    - filtering of diagnostic messages a la (java) commonLogging
      implemented.

0.05  Wed May 10 2006
	- Rebuild of MANIFEST - some newly added files were missing.

0.04  Fri May 5 2006	Tony Byrne
	- Added an 'abstract' for search.cpan.org

0.03  Tue May 2 2006	Tony Byrne
	- Release to include missing files which had been submitted by Martin.

0.02  Mon May 1 2006    Martin Busik <martin.busik@busik.de>
	- simple entry point for FileRunner and WikiRunner
		perl -MTest::C2FIT -e file_runner <input-file> <output-file>
	- Output and result code of FileRunner and WikiRunner identical to
		the reference implementation (java implementation, version 1.1).
	- loading of Fixtures/Packages modified.
		The specification requires the usage of java-style package
		separators ("." instead of the perl-style "::"). Further, the
		(input) documents shall use fully qualified package names. In
		order to allow the usage of at least the specification documents
		without change, special care is taken of the "fit.*" names.
		Similar rules apply to "eg.*" and "fat.*" names.
		The loader does not require each referenced package to be in a
		separate file now.
	- Added better support for TypeAdapters.
		Test::C2FIT::Fixture implements the following methods which,
		given a field or method name, returns the (fully qualified) package
		name of the appropriate TypeAdapter:
		suggestFieldType - used in ColumnFixture an RowFixture
		suggestMethodResultType - used in ColumnFixture, RowFixture and 
			ActionFixture
		suggestMethodParamType - used in ActionFixture
		by default, these methods make a lookup in an instance's hash. So 
		either these methods are to overload or the appropriate hashes 
		are to be filled.
	- ActionFixture
		loading of fixtures (see above)
		added TypeAdapter-support in do_enter
	- ColumnFixture
		Bug in check() fixed. When execute() was called within the check()
		method and execute() raised an exception, further handling ignored
		this. (This bug is still exists in V1.1 of the java implementation)
	- Fixture
		TypeAdapter support (see above)
		Bug in doRows fixed. Wrong iterator implementation, causing endless
		loops if doRow() added further rows to the output.
		implementation of exception() now identical to the java impl.
		Implementation of check() regarding empty cells modified, behavior
		now identical to the java impl.
		Error text in loadFixture() modified according to the specification
		(less verbose)
	- GenericArrayAdapter
		implementation of equals modified.
	- Parse
		Empty tags, e.g. "<td/>" will be now correctly parsed. Until yet,
		there were an exeption (same problem occurs in the V1.1 of the
		reference implementation)
		Added dummy implementation of footnote()
	- RowFixture
		Printout of surplus rows corrected.
	- TimedActionFixture
		Time is printed in human readable form now
	- TypeAdapter
		Added better support for (specific) TypeAdapters
		Implmenentation of equals() changed. In java, equals() is a method
		of Object, it can be generally used. This is not required in perl,
		but needed, when (propably) blessed objects are in place (e.g.
	    { a => 1, b => 2 } eq { a => 1, b => 2} is not true).
		The current implementation accomodates this.
		For numeric comparison, the ScientificDouble is used
		(as this is the lesser evil than abs($a - b) < 0.00001 :)
		Given an undef, toString return now "null" (as in the V1.1 of
		the reference implementation)
	- Generally: Behavior checked against the version 1.1 of the java
	  	implementation. 
	- All examples of the reference implementation now also included.
		The results are as identical to the reference implementation as
		possible (e.g. "integer overflow")
	- MINOR changes in the examples/input/*.html files.
		Some files contained path specifiations with wildcards
		(e.g. "Documents/AllFiles/magnitude/*.html) In order to get these
		running, "Documents" was replaced by "input".
		The WebPageExample refers now to URLs which (at present: 2006-05-01)
		exist.
	- examples-perl added
		Contains an example which uses the above mentioned TypeAdapter 
		features.

0.01  Wed Apr 27 17:10:52 2005
	- Initial version.



run_tests.pl
#!/usr/bin/perl -w

# $Id: run_tests.pl,v 1.6 2006/05/15 08:37:07 tonyb Exp $
#
# Copyright (c) 2002-2005 Cunningham & Cunningham, Inc.
# Released under the terms of the GNU General Public License version 2 or later.
#
# Written by Tony Byrne <fit4perl@byrnehq.com>

use strict;
use lib qw( lib t/lib examples/lib );
use Test::Unit::TestRunner;

my @tests = @ARGV ? @ARGV : 'Test::C2FIT::test::AllTests';
my $testrunner = Test::Unit::TestRunner->new();

while (my $test = shift @tests)
{
	my $package = &pathToPackageName($test);
	$testrunner->start($test);
}

sub pathToPackageName
{
	my $path = shift;
	my $package = $path;
	$package =~ s/^t\/lib\///;
	$package =~ s/\.pm$//g;
	$package =~ s/\//::/g;
	return $package;
}

lib/Test/C2FIT.pm
package Test::C2FIT;

#use 5.008006;
use Test::C2FIT::FileRunner;
use Test::C2FIT::WikiRunner;
use Exporter();
use Getopt::Std;
@ISA    = qw(Exporter);
@EXPORT = qw(file_runner wiki_runner fit_shell);
use strict;
use warnings;

our $VERSION = '0.07';

sub file_runner {
    my $param = {};
    die "unsupported param!" unless getopt( "L:", $param );
    local $SIG{'__WARN__'} = _commonLogging( $param->{L} );

    unshift( @INC, '.' ) unless grep { /^\.$/ } @INC;
    Test::C2FIT::FileRunner->new()->run(@ARGV);
}

sub wiki_runner {
    my $param = {};
    die "unsupported param!" unless getopt( "L:", $param );
    local $SIG{'__WARN__'} = _commonLogging( $param->{L} );

    unshift( @INC, '.' ) unless grep { /^\.$/ } @INC;
    Test::C2FIT::WikiRunner->new()->run(@ARGV);
}

sub fit_shell {
    my $shell = Test::C2FIT::_Shell->new();
    $shell->init;
    $shell->runShell(*STDIN);
}

sub _commonLogging {
    my $loglevel = shift;
    $loglevel = 3 unless defined($loglevel);

    return sub {
        local ( $_, $&, $1 );

        if ( defined( $_[0] ) && $_[0] =~ /^(\d+)/ ) {
            return unless $1 >= $loglevel;
        }

# my ($sec,$min,$hour,$mday,$mon,$year,undef,undef,undef) = localtime(time);
# $mon++;
# $year = $year + 1900;
# print STDERR sprintf("%02d.%02d.%04d %02d:%02d:%02d ",$mday,$mon,$year,$hour,$min,$sec);
        print STDERR @_;
    };
}

#
#   (help-) Package implementing an interactive shell
#
{

    package Test::C2FIT::_Shell;
    use Config;
    use strict;
    use warnings;

    sub new {
        my $pkg  = shift;
        my $self = {
            lastCmd    => undef,
            perlBinary => $Config{perlpath},
            options    => {
                input     => ".",
                output    => ".",
                inc       => ".",
                perl_opt  => undef,
                log_level => undef,
                runner    => 'file_runner',
                verbose   => 0,
            },
            cmdDispatch => {
                help   => \&Test::C2FIT::_Shell::help,
                show   => \&Test::C2FIT::_Shell::show,
                set    => \&Test::C2FIT::_Shell::set,
                run    => \&Test::C2FIT::_Shell::run,
                runall => \&Test::C2FIT::_Shell::runall,
                nop    => sub { 1 },
                quit   => sub { undef },
            },
            parseRunDispatch => {
                rerun  => \&Test::C2FIT::_Shell::rerun,
                infile => \&Test::C2FIT::_Shell::runInfile,
                files  => \&Test::C2FIT::_Shell::runFiles,
                inout  => \&Test::C2FIT::_Shell::runInOut,
            }
        };
        return bless $self, $pkg;
    }

    sub init {
        my $self = shift;

        #
        #   check if input/output/lib exists
        #
        my $v = {};

        $v->{input}  = "input"  if ( -d "./input" );
        $v->{output} = "output" if ( -d "./output" );
        $v->{inc}    = "lib"    if ( -d "./lib" );

        if ( scalar(%$v) ) {
            while ( my ( $ik, $iv ) = each(%$v) ) {
                $self->{options}->{$ik} = $iv;
            }
            $self->msg(
                "INFO: input/ouptut/lib diretories found, using them!\n");
            $self->show;
        }
    }

    sub help {
        print <<'_EOH_'; return 1;
# Supported commands:
# help      - show this help
# show      - show variables
# set <variable> <value> - set operational variables
#   input                - directory where html files are searched
#   output               - directory where result html will be written
#   inc                  - INC path. This will be added to the perl-process
#                          upon creation (perl -Ip1 -Ip2 etc.)
#                          multiple Entries should be separated either by
#                          colon (:) or by semicolon, e.g. lib:.:../test
#                          if @ is given, then then contents of @INC of the
#                          fit_shell-process will be set
#   perl_opt             - optional perl-parameters to the perl process.
#                          e.g. -d for debugging
#   log_level            - will be passed to the runner
#   runner               - either file_runner or wiki_runner
#   verbose              - either 0 or 1. When 1, then the command actually
#                          run will be printed too
# run      - run a document. There are different ways to specify it:
#   run BinaryChop       - runs a file named $input/BinaryChop.html
#                          output goes to $output/BinaryChop.html
#   run BinaryChop out   - runs a file named $input/BinaryChop 
#                          (without extension!!!)
#                          output goes to $output/out (without extension!)
#
#   run                  - rerun last run command
#
#   run *.htm*           - run all files $input/*.htm*
#                          (same as runall)
#   run a*.html b*.html  - run all files $input/a*.html $input/b*.html
#                          (Difference to the other two-Param run-call: 
#                           here, wildcards are used)
#   runall               - same as run *.htm*
#
#   quit                 - terminate this shell
#   
#   For each run, a new perl process is started.
#
#   There are some shortcuts/aliases too:
#
#   !                   is an alias for "run"
#   [EMPTY LINE]        (i.e. just pressing the enter key) rerun last run
#
#   Lines starting with # will be ignored
#
#   All output of the fit_shell starts with "#", so you can easily eliminate
#   it (e.g. grep -v "^#")
#
_EOH_
    }

    sub msg {
        my $self = shift;
        print "# ", @_;
        1;
    }

    sub show {
        my $self = shift;
        my @k    = sort keys %{ $self->{options} };

        for my $k (@k) {
            my $v = $self->{options}->{$k};
            $v = "<not defined>" unless defined($v);
            print sprintf( "# %10s: %s\n", $k, $v );
        }
        1;
    }

    sub set {
        my ( $self, $rest ) = @_;
        return $self->msg("WARN: wrong syntax for set!\n")
          unless $rest =~ /(\S+)(?:\s+(\S.*))?$/;

        my $key = $1;
        my $val = $2;

        return $self->msg("WARN: invalid variable $key!\n")
          unless exists $self->{options}->{$key};

        if ( $key eq "inc" && $val eq "@" ) {
            $val = join( ":", @INC );
        }

        $self->{options}->{$key} = $val;
        $self->{lastCmd} = undef;
        1;
    }

    sub runall {
        my $self = shift;
        $self->run("*.htm*");
    }

    sub run {
        my ( $self, $rest ) = @_;
        my $dispatch = $self->{parseRunDispatch};

        my ( $key, @vals ) = $self->parseRunCmd($rest);
        die "internal error. unknown state: $key\n"
          unless exists $dispatch->{$key};

        my $code = $dispatch->{$key};
        return $code->( $self, @vals );
    }

    sub rerun {
        my $self = shift;
        return $self->msg("WARN: no cmd to rerun!\n")
          unless defined( $self->{lastCmd} );
        return $self->_run( $self->{lastCmd} );
    }

    #
    #   runInOut($in,$out) - both filenames with path and suffix
    #
    sub runInOut {
        my ( $self, $in, $out ) = @_;

        return $self->msg("WARN: in and out identical: $in Will be ignored!\n'")
          if $in eq $out;

        my $cmd = $self->_buildCmd;
        $cmd .= " $in $out";
        return $self->_run($cmd);
    }

    #
    #   runInfile (path+filename) * optionally without the .html suffix
    #
    sub runInfile {
        my ( $self, $in ) = @_;
        my $input  = quotemeta( $self->{options}->{input} );
        my $output = $self->{options}->{output};

        $in .= ".html" unless $in =~ /\.html$/i;
        my $out = $in;
        $out =~ s/^$input/$output/;

        return $self->runInOut( $in, $out );
    }

    #
    #   runFiles(@files) - list of input file names with path and extension
    #
    sub runFiles {
        my ( $self, @files ) = @_;
        for my $f (@files) {
            $self->runInfile($f);
        }
        1;
    }

    sub _run {
        my ( $self, $cmd ) = @_;
        die "no cmd given!" unless defined($cmd);
        $self->{lastCmd} = $cmd;
        $self->msg("# CMD:$cmd\n") if $self->{options}->{verbose};
        system($cmd);
        1;
    }

    sub _buildCmd {    # setup the command up to ARGV
        my $self    = shift;
        my $cmd     = $self->{perlBinary};
        my $options = $self->{options};

        $cmd .= " " . $options->{perl_opt} . " "
          if defined( $options->{perl_opt} );

        if ( defined( $options->{inc} ) ) {
            my @inc = split( /[:;]/, $options->{inc} );
            my $inc = " -I" . join( " -I", @inc );
            $cmd .= $inc;
        }

        $cmd .= " -MTest::C2FIT -e " . $options->{runner} . " -- ";

        my $logLevel = $options->{log_level};
        $cmd .= " -L $logLevel " if defined($logLevel);

        return $cmd;
    }

    sub parseRunCmd {
        my ( $self, $rest ) = @_;
        my $input        = $self->{options}->{input};
        my $output       = $self->{options}->{output};
        my $hasWildcards = 0;
        my @files        = ();

        return ("rerun") unless defined($rest);

        my @items = map { /\*/ && $hasWildcards++; $_ } split /\s+/, $rest;

        if ($hasWildcards) {
            for my $item (@items) {
                push( @files, glob("$input/$item") );
            }
            return ( "files", @files );
        }
        if ( 2 == @items ) {
            return ( "inout", "$input/$items[0]", "$output/$items[1]" );
        }
        return ( "infile", "$input/$items[0]" );
    }

    sub runShell {    # main loop of fit_shell
        my ( $self, $inFH ) = @_;
        my $line;
        my $dispatch = $self->{cmdDispatch};
        my $prompt   = ( -t "$inFH" ) ? "fit> " : "";

        $| = 1;

        print $prompt;
        while ( $line = <$inFH> ) {
            $line =~ s/(\012\015?|\015\012?)$//;
            my ( $cmd, $rest ) = $self->parseCmd($line);

            die "internal error. Unknown command/state: $cmd\n"
              unless exists $dispatch->{$cmd};

            my $code = $dispatch->{$cmd};
            my $rv = $code->( $self, $rest );
            last unless $rv;
            print $prompt;
        }
    }

    sub parseCmd {
        my ( $self, $line ) = @_;

        return ( "run", undef ) if !defined($line) || $line eq "";
        return ("nop") if $line =~ /^\s*#/;

        $line =~ s/^\s+//;
        $line =~ s/\s+$//;

        if ( $line =~ /^!\s*(\S.*)?$/ ) {    # ! is alias for "run"
            return ( "run", $1 );
        }
        if ( $line =~ /^(\w+)\b\s*(\S.*)?$/ ) {
            return ( $1, $2 );
        }
        return ("nop");
    }

    1;
};

1;

__END__

=head1 NAME

Test::C2FIT - A direct Perl port of Ward Cunningham's FIT 
acceptance test framework for Java.

=head1 VERSION

This document describes Test::C2FIT version 0.06

=head1 SYNOPSIS

	FileRunner.pl input_containing_fit_tests.html test_results.html

	perl -MTest::C2FIT -e file_runner input_containing_fit_tests.html test_results.html

	perl -MTest::C2FIT -e fit_shell

=head1 DESCRIPTION

Great software requires collaboration and communication. Fit is a tool for
enhancing collaboration in software development. It's an invaluable way
to collaborate on complicated problems - and get them right - early
in development.

Fit allows customers, testers, and programmers to learn what their 
software should do and what it does do. It automatically compares
customers' expectations to actual results.

This port of FIT has a featureset equivalent to v1.1 of FIT.  
Dave W. Smith's original port was based on fit-b021021j and I've updated 
most of the core to match the 1.1 version.

This port passes the current FIT spec and also implements a all of the
examples.


The following functions are provided (and exported) by this module:

=over 4

=item B<file_runner($infile,$outfile)>

Process a FIT-document contained in $infile and writes the result to $outfile.

=item B<wiki_runer($infile,$outfile)>

Same as file_runner, except that not E<lt>tableE<gt>, E<lt>trE<gt> and E<lt>tdE<gt>
but E<lt>wikiE<gt>, E<lt>tableE<gt>, E<lt>trE<gt> and E<lt>tdE<gt> is searched for in the input
document.

=item B<fit_shell>

Creates an interactive shell from which you can easily run tests. Start it
and enter "help" for more information.

Suppose, your tests-related files reside in a directory with three subdirectories:
input - where the files come from, output - where the results will be written to and
lib - where your fixtures reside, all you need to do is just to enter "runall"

=back


=head2 Logging

The file_runner and wiki_runner support filtering of warn messages, similar to
java's common logging. To change the log level, use the -L parameter, e.g.:

	perl -MTest::C2FIT -e file_runner -- -L 1 input_containing_fit_tests.html test_results.html

There are following log levels defined: 0 - trace, 1 - debug, 2 - info, 3 - warn,
4 - error, 5 - fatal.

In your code, simply use C<warn "message"> if it should be printed out unconditionally or
C<warn 1, " message"> if it should be printed out, when log level is either TRACE or DEBUG.


=head2 Naming, Namespace(s)

In your FIT-documents, please use the java-style dot-notation for qualifying package names. E.g.
if you want the package Domain::Object::Simple to be used, specify it by entering 
Domain.Object.Simple into your fit document.


Package names should be fully qualified, case is importat.
Special care is taken on the fit.* packages, these can be specified either by fit.Name as
well as Test.C2FIT.Name.


=head1 GOTCHAS AND LIMITATIONS

1) Java is a strongly typed language; Perl is not. The Java version of FIT
cares a lot about types, but Perl takes a more relaxed view of things and
this port reflects that.

2) Perl supports limited introspection. Because there are no method signatures,
it isn't possible to determine method return types. If you want to use
TypeAdapters you have to supply hints. (see examples)

3) Some of the tests from the 'examples' directory expect Java behaviour for
arithmetic (e.g. integer overflow).  Perl doesn't have this type of overflow
so these tests will "fail".

4) The MusicExample uses a clock that doesn't have millisecond accuracy. This
throws off the clock by a second during one of the tests.

5) Perl supports a limited set of primitive types. Dave has used a
GenericTypeAdapter that knows about strings and numbers (and pretends
to know about booleans).


=head1 SEE ALSO

Extensive and up-to-date documentation on FIT can be found at:
http://fit.c2.com/

The 'examples' directory of this distribution contains some sample FIT
tests and sample applications that they test.  Invoke FileRunner.pl on
any of the test input files from examples/input and view the output in
a browser. To invoke the tests use do_tests.bat / do_tests.sh in
the appropriate directory.

The directory examples-perl contains examples written for this perl-port
only.

You should also examine and run the tests in the 'spec' directory.
These are FIT's own acceptance tests.

=head1 AUTHOR

Original port from the Java version by Dave W. Smith.

Updates and modifications by Tony Byrne E<lt>fit4perl@byrnehq.comE<gt>.
Further modifications by Martin Busik E<lt>martin.busik@busik.deE<gt>.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2002-2006 Cunningham & Cunningham, Inc.
Released under the terms of the GNU General Public License version 2 or later.


=cut


examples/run_tests.sh
#!/bin/sh

perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/AllCombinations.html output/AllCombinations.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/AllFiles.html output/AllFiles.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/AllPairs.html output/AllPairs.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/BasicEmployeeCompensation.html output/BasicEmployeeCompensation.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/BinaryChop.html output/BinaryChop.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/CalculatorExample.html output/CalculatorExample.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/ExampleTests.html output/ExampleTests.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/FitAcceptanceTests.html output/FitAcceptanceTests.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/MusicExample.html output/MusicExample.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/MusicExampleWithErrors.html output/MusicExampleWithErrors.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/NetworkExample.html output/NetworkExample.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/WebPageExample.html output/WebPageExample.html
perl -I../lib -Ilib -MTest::C2FIT -e file_runner input/arithmetic.html output/arithmetic.html



examples-perl/lib/RowsetFilter.pm
#
#   Martin Busik <martin.busik@busik.de>
#
#
#

package RowsetFilter;
use base qw(Test::C2FIT::Fixture);

use vars qw($filteredRows);

sub new {
    my $pkg = shift;
    my $types = { findByManagerId => 'NullFkWrapper',
                  findById => 'NullFkWrapper' };

    $filteredRows = [] unless defined($filteredRows);

    return $pkg->SUPER::new(@_,methodSetterTypeMap => $types);
}

sub clear {
    # no data in output
    $filteredRows = [];
}

sub allRows {
    $filteredRows = $NullValueTest::ourData;
}

sub rowCount {
    my $self = shift;
    return scalar(@$filteredRows)
}

sub findById {
    my ($self,$id) = @_;
    $self->doFilter(filterByFieldname('id',$id));
}

sub findByManagerId {
    my ($self,$managerId) = @_;
    $self->doFilter(filterByFieldname('manager_id',$managerId));
}

sub filterByFieldname {
    my ($fieldname,$value) = @_;

    return (defined($value)) ?
        sub { $_[0]->{$fieldname} == $value } :
        sub { !defined($_[0]->{$fieldname}) };
}

sub doFilter {
    my ($filter,$filterCode) = @_;

    my @newRows = grep { $filterCode->($_) } @$NullValueTest::ourData;
    $filteredRows = \@newRows;
}

1;


Makefile.PL
#use 5.008006;
use ExtUtils::MakeMaker;

{
package MY;

sub metafile {
    my $self = shift;
    my $inject = <<"META_EXTRA";
abstract: $self->{ABSTRACT}
author: 
  - $self->{AUTHOR}
no_index:
  dir:
    - t
    - examples
    - examples-perl
    - spec
    - html
meta-spec:
  version: 1.2
  url: http://module-build.sourceforge.net/META-spec-v1.2.html
resources:
  homepage: http://fit.c2.com 
META_EXTRA

    $inject = join "\n\t",  map { '$(NOECHO) $(ECHO) "'.$_.'" >>META_new.yml' }
                            split /(?:\012\015?|\015\012?)/, $inject;

    my $t = $self->SUPER::metafile_target(@_);
    my $mvCmd = q|-$(NOECHO) $(MV)|;
    my $pat = quotemeta($mvCmd);
    $t =~ s/$pat/$inject\n\t$mvCmd/s;
    return $t;
}

sub postamble {
   	q ( 
all :: ReleaseNotes

# Make search.cpan.org happy but still follow GNU standards:
# (Thanks to Graham Barr for the hint)
ReleaseNotes: Changes
	$(ABSPERLRUN) -MExtUtils::Command -e cat Changes >$@
	

# simple pod 2 html conversion
my_html :
	$(ABSPERLRUN) html/html.pl html

);
}

1;
};

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Test::C2FIT',
    VERSION_FROM      => 'lib/Test/C2FIT.pm', # finds $VERSION
    PREREQ_PM         => {
				Test::Unit => 0.24
			}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT       => "A direct Perl port of Ward Cunningham's FIT acceptance test framework for Java",
       AUTHOR         => 'Tony Byrne <fit4perl@byrnehq.com>') : ()),
    # LICENSE => ???,
);
